File: mid/Graph/133. Clone Graph (DFS+Hash).py
# https://leetcode.com/problems/clone-graph/solutions/3393494/python-simple-and-clean-beats-94-55/
# Time Complexity : O(N), where N is the number of nodes in the graph and E is the number of edges in the graph.
# Space Complexity : O(N), where N is the number of nodes in the graph. 
# DFS + Stack
# depth-first search (DFS) algorithm to traverse the graph and create the deep copy.

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        mapping = {}
        stack = [node]
        while stack:
            curr = stack.pop()
            if curr.val not in mapping:
                mapping[curr.val] = Node(curr.val)
            if curr.neighbors != []:
                for neighbor in curr.neighbors:
                    if neighbor.val not in mapping:
                        mapping[neighbor.val] = Node(neighbor.val)
                        stack.append(neighbor)
                    mapping[curr.val].neighbors.append(mapping[neighbor.val])
        return mapping[node.val]

File: mid/Graph/133. Clone Graph (DFS+Hash).cpp
// https://leetcode.com/problems/clone-graph/solutions/3393172/day-98-dfs-hash-table-easiest-beginner-friendly-sol/
// Time Complexity : O(N + E), where N is the number of nodes in the graph and E is the number of edges in the graph.
// Space Complexity : O(N), where N is the number of nodes in the graph. 
// DFS + Hash Table

class Solution {
public:
    unordered_map<Node*, Node*> cloningGraph;
    
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return nullptr;
        }
        if (cloningGraph.find(node) == cloningGraph.end()) {
            cloningGraph[node] = new Node(node -> val);
            for (auto neighbor : node -> neighbors) {
                cloningGraph[node] -> neighbors.push_back(cloneGraph(neighbor));
            }
        }
        return cloningGraph[node];
    }
};